Information about all the attributes can be found here:

https://www.kaggle.com/c/allstate-claims-severity/data

The aim of the challenge is to predict the 'loss' based on the variables in the dataset. Hence, this is a regression problem.

In [28]:
# Read raw data from the file



import numpy as np
import pandas as pd
from pandas import Series,DataFrame

from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

import seaborn as sns

import matplotlib.pyplot as plt
%matplotlib inline

import os,sys
from scipy import stats


Read the train dataset

In [34]:
allstate = pd.read_csv('train.csv',skipinitialspace=True)

Display the first five rows of the data

In [35]:
allstate.head()

id cat1 cat2 cat3 cat4 cat5 cat6 cat7 cat8 cat9   ...        cont6  \
0   1    A    B    A    B    A    A    A    A    B   ...     0.718367   
1   2    A    B    A    A    A    A    A    A    B   ...     0.438917   
2   5    A    B    A    A    B    A    A    A    B   ...     0.289648   
3  10    B    B    A    B    A    A    A    A    B   ...     0.440945   
4  11    A    B    A    B    A    A    A    A    B   ...     0.178193   

      cont7    cont8    cont9   cont10    cont11    cont12    cont13  \
0  0.335060  0.30260  0.67135  0.83510  0.569745  0.594646  0.822493   
1  0.436585  0.60087  0.35127  0.43919  0.338312  0.366307  0.611431   
2  0.315545  0.27320  0.26076  0.32446  0.381398  0.373424  0.195709   
3  0.391128  0.31796  0.32128  0.44467  0.327915  0.321570  0.605077   
4  0.247408  0.24564  0.22089  0.21230  0.204687  0.202213  0.246011   

     cont14     loss  
0  0.714843  2213.18  
1  0.304496  1283.60  
2  0.774425  3005.09  
3  0.602642   939.85  
4  0.432606  2763.85  

[5 rows x 132 columns]


Data statistics : Description

In [39]:
allstate.describe()

id           cat1           cat2           cat3  \
count  188318.000000  188318.000000  188318.000000  188318.000000   
mean   294135.982561       0.248346       0.433294       0.054827   
std    169336.084867       0.432055       0.495532       0.227644   
min         1.000000       0.000000       0.000000       0.000000   
25%    147748.250000       0.000000       0.000000       0.000000   
50%    294539.500000       0.000000       0.000000       0.000000   
75%    440680.500000       0.000000       1.000000       0.000000   
max    587633.000000       1.000000       1.000000       1.000000   

                cat4           cat5           cat6           cat7  \
count  188318.000000  188318.000000  188318.000000  188318.000000   
mean        0.318201       0.342936       0.300688       0.024289   
std         0.465779       0.474692       0.458559       0.153944   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         1.000000       1.000000       1.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

                cat8           cat9      ...                cont6  \
count  188318.000000  188318.000000      ...        188318.000000   
mean        0.058645       0.399303      ...             0.490945   
std         0.234961       0.489757      ...             0.205273   
min         0.000000       0.000000      ...             0.012683   
25%         0.000000       0.000000      ...             0.336105   
50%         0.000000       0.000000      ...             0.440945   
75%         0.000000       1.000000      ...             0.655021   
max         1.000000       1.000000      ...             0.997162   

               cont7          cont8          cont9         cont10  \
count  188318.000000  188318.000000  188318.000000  188318.000000   
mean        0.484970       0.486437       0.485506       0.498066   
std         0.178450       0.199370       0.181660       0.185877   
min         0.069503       0.236880       0.000080       0.000000   
25%         0.350175       0.312800       0.358970       0.364580   
50%         0.438285       0.441060       0.441450       0.461190   
75%         0.591045       0.623580       0.566820       0.614590   
max         1.000000       0.980200       0.995400       0.994980   

              cont11         cont12         cont13         cont14  \
count  188318.000000  188318.000000  188318.000000  188318.000000   
mean        0.493511       0.493150       0.493138       0.495717   
std         0.209737       0.209427       0.212777       0.222488   
min         0.035321       0.036232       0.000228       0.179722   
25%         0.310961       0.311661       0.315758       0.294610   
50%         0.457203       0.462286       0.363547       0.407403   
75%         0.678924       0.675759       0.689974       0.724623   
max         0.998742       0.998484       0.988494       0.844848   

                loss  
count  188318.000000  
mean     3037.337686  
std      2904.086186  
min         0.670000  
25%      1204.460000  
50%      2115.570000  
75%      3864.045000  
max    121012.250000  

[8 rows x 132 columns]

  Machine Learning Algorithms : Evaluation and Prediction

  Linear Regression 

In [40]:

# Import for Linear Regression
import sklearn
from sklearn.linear_model import LinearRegression

# Create a LinearRegression Object
lreg = LinearRegression()



Data Preparation: Categorical Variables

In [41]:
# Binarizing Categorical Variables

import pandas as pd
features = allstate.columns
cats = [feat for feat in features if 'cat' in feat]
for feat in cats:
    allstate[feat] = pd.factorize(allstate[feat], sort=True)[0]

Splitting the data into train and test

In [42]:
# Preparing data for train and test split 
from sklearn.cross_validation import train_test_split
from sklearn.metrics import accuracy_score, mean_absolute_error
import numpy 

X=allstate.drop(['id','loss'],1).fillna(value=0)
Y=allstate['loss']
X.head()


seed = 8
test_size = 0.33
X_train, X_test, y_train, y_test = train_test_split(X,Y, test_size=test_size, random_state=seed)


cat1  cat2  cat3  cat4  cat5  cat6  cat7  cat8  cat9  cat10    ...     \
0     0     1     0     1     0     0     0     0     1      0    ...      
1     0     1     0     0     0     0     0     0     1      1    ...      
2     0     1     0     0     1     0     0     0     1      1    ...      
3     1     1     0     1     0     0     0     0     1      0    ...      
4     0     1     0     1     0     0     0     0     1      1    ...      

      cont5     cont6     cont7    cont8    cont9   cont10    cont11  \
0  0.310061  0.718367  0.335060  0.30260  0.67135  0.83510  0.569745   
1  0.885834  0.438917  0.436585  0.60087  0.35127  0.43919  0.338312   
2  0.397069  0.289648  0.315545  0.27320  0.26076  0.32446  0.381398   
3  0.422268  0.440945  0.391128  0.31796  0.32128  0.44467  0.327915   
4  0.704268  0.178193  0.247408  0.24564  0.22089  0.21230  0.204687   

     cont12    cont13    cont14  
0  0.594646  0.822493  0.714843  
1  0.366307  0.611431  0.304496  
2  0.373424  0.195709  0.774425  
3  0.321570  0.605077  0.602642  
4  0.202213  0.246011  0.432606  

[5 rows x 130 columns]

In [43]:

# Implement Linear Regression

lreg.fit(X_train,y_train)       

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [44]:
# Predictions on training and testing sets
pred_train = lreg.predict(X_train)
pred_test = lreg.predict(X_test)

Accuracy of the model :Linear regression

In [45]:

from sklearn import metrics
print('MAE:', metrics.mean_absolute_error(y_test, pred_test))
print('MSE:', metrics.mean_squared_error(y_test, pred_test))
print('RMSE:', np.sqrt(metrics.mean_squared_error(y_test, pred_test)))

MAE: 1331.21132688
MSE: 4328206.19396
RMSE: 2080.43413593



  Machine learning algorithm : Ridge Regression for Evaluation, prediction


In [46]:
# Implement Ridge Regression

from sklearn import linear_model
reg = linear_model.Ridge (alpha = 0.1)
reg.fit (X_train,y_train)

Ridge(alpha=0.1, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)

In [47]:
# Predictions on training and testing sets
pred_train = reg.predict(X_train)
pred_test = reg.predict(X_test)


from sklearn import metrics
print('MAE:', metrics.mean_absolute_error(y_test, pred_test))
print('MSE:', metrics.mean_squared_error(y_test, pred_test))
print('RMSE:', np.sqrt(metrics.mean_squared_error(y_test, pred_test)))

MAE: 1331.23071391
MSE: 4328239.06277
RMSE: 2080.44203543


 Machine learning algorithm : Lasso Regression for Evaluation, prediction


In [50]:
 from sklearn import linear_model
reg = linear_model.Lasso(alpha = 0.1)
reg.fit (X_train,y_train)

pred_test = reg.predict(X_test)


from sklearn import metrics
print('MAE:', metrics.mean_absolute_error(y_test, pred_test))
print('MSE:', metrics.mean_squared_error(y_test, pred_test))
print('RMSE:', np.sqrt(metrics.mean_squared_error(y_test, pred_test)))


/usr/local/lib/python3.5/dist-packages/sklearn/linear_model/coordinate_descent.py:484: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)


Lasso(alpha=0.1, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False)

MAE: 1330.93688949
MSE: 4327206.58149
RMSE: 2080.19388074


 Machine learning algorithm : KNN for Evaluation, prediction


In [16]:
# loading library
from sklearn.neighbors import KNeighborsRegressor

# instantiate learning model (k = 3)
knn = KNeighborsRegressor(n_neighbors=3)

# fitting the model
knn.fit(X_train, y_train)

# predict the response
pred = knn.predict(X_test)



from sklearn import metrics
print('MAE:', metrics.mean_absolute_error(y_test, pred))
print('MSE:', metrics.mean_squared_error(y_test, pred))
print('RMSE:', np.sqrt(metrics.mean_squared_error(y_test, pred)))

MAE: 1715.26231722
MSE: 7444923.32293
RMSE: 2728.53867902


 Machine learning algorithm : Decisiontrees for Evaluation, prediction

In [49]:
from sklearn import tree
model = tree.DecisionTreeRegressor()
model.fit(X_train, y_train)
# Predictions on training and testing sets
pred_train = model.predict(X_train)
pred_test = model.predict(X_test)

from sklearn import metrics
print('MAE:', metrics.mean_absolute_error(y_test, pred_test))
print('MSE:', metrics.mean_squared_error(y_test, pred_test))
print('RMSE:', np.sqrt(metrics.mean_squared_error(y_test, pred_test)))

DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
           max_leaf_nodes=None, min_impurity_split=1e-07,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, presort=False, random_state=None,
           splitter='best')

MAE: 1745.6449576
MSE: 7751826.99311
RMSE: 2784.21029973


Decisiontrees for Evaluation, prediction with cross validation

In [18]:

from sklearn import tree
import numpy as np
import pandas as pd
from sklearn.cross_validation import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import accuracy_score
from sklearn import tree

clf_gini = DecisionTreeRegressor(random_state = 100,
                               max_depth=5, min_samples_leaf=2)
clf_gini.fit(X_train, y_train)

# Predictions on training and testing sets
pred_train = clf_gini.predict(X_train)
pred_test = clf_gini.predict(X_test)

from sklearn import metrics
print('MAE:', metrics.mean_absolute_error(y_test, pred_test))
print('MSE:', metrics.mean_squared_error(y_test, pred_test))
print('RMSE:', np.sqrt(metrics.mean_squared_error(y_test, pred_test)))

MAE: 1454.94423059
MSE: 5080915.67358
RMSE: 2254.08865699


Decisiontrees for Evaluation, prediction with cross validation and hyperparameter tuning

In [19]:
from sklearn.tree import DecisionTreeRegressor
from sklearn import metrics
from sklearn import grid_search
param_grid = {'max_depth': np.arange(3, 10)}

tree = grid_search.GridSearchCV(DecisionTreeRegressor(), param_grid)

tree.fit(X_train, y_train)
pred_test = tree.predict(X_test)
tree_performance = metrics.mean_absolute_error(y_test, pred_test)

print (format(tree_performance))

/usr/local/lib/python3.5/dist-packages/sklearn/grid_search.py:43: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. This module will be removed in 0.20.
  DeprecationWarning)


1378.3208319624282


Randomforest  for Evaluation, prediction

In [20]:

#random forest
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor(n_estimators=5)
model.fit(X_train, y_train)
# Predictions on training and testing sets
pred_train = model.predict(X_train)
pred_test = model.predict(X_test)

from sklearn import metrics
print('MAE:', metrics.mean_absolute_error(y_test, pred_test))
print('MSE:', metrics.mean_squared_error(y_test, pred_test))
print('RMSE:', np.sqrt(metrics.mean_squared_error(y_test, pred_test)))


MAE: 1375.4761736
MSE: 4609842.82506
RMSE: 2147.05445321


In [22]:
#grid search cv for best params
#couldnt do it for computational reasons


#n_features = X.shape[1]

#from sklearn.grid_search import RandomizedSearchCV
#grid = RandomizedSearchCV(model, n_iter=20,
 #           param_distributions=dict(
  #                                        max_depth=np.arange(5,20+1),
   #                                       max_features=np.arange(1, n_features+1)
    #                                )
    #     )
#grid.fit(X, Y)
#print(grid.best_params_)
#
#model = RandomForestRegressor(max_features=grid.best_params_["max_features"],
 #                             max_depth=grid.best_params_["max_depth"])
#model.fit(train_X, train_y)
# Predictions on training and testing sets
#pred_train = model.predict(X_train)
#pred_test = model.predict(X_test)

#from sklearn import metrics
#print('MAE:', metrics.mean_absolute_error(y_test, pred_test))
#print('MSE:', metrics.mean_squared_error(y_test, pred_test))
#print('RMSE:', np.sqrt(metrics.mean_squared_error(y_test, pred_test)))


Adaboost regression for Evaluation, prediction


In [11]:
# importing necessary libraries
import numpy as np
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import AdaBoostRegressor


In [16]:

regr = AdaBoostRegressor(DecisionTreeRegressor(max_depth=4),
                          n_estimators=300, random_state = 100)

regr.fit(X_train, y_train)

AdaBoostRegressor(base_estimator=DecisionTreeRegressor(criterion='mse', max_depth=4, max_features=None,
           max_leaf_nodes=None, min_impurity_split=1e-07,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, presort=False, random_state=None,
           splitter='best'),
         learning_rate=1.0, loss='linear', n_estimators=300,
         random_state=100)

In [17]:
# Predictions on training and testing sets
pred_train = regr.predict(X_train)
pred_test = regr.predict(X_test)

from sklearn import metrics
print('MAE:', metrics.mean_absolute_error(y_test, pred_test))
print('MSE:', metrics.mean_squared_error(y_test, pred_test))
print('RMSE:', np.sqrt(metrics.mean_squared_error(y_test, pred_test)))

MAE: 6840.07247307
MSE: 50980307.3071
RMSE: 7140.04953114


Gradient boosting trees regression for Evaluation, prediction


In [18]:
from sklearn.ensemble import GradientBoostingRegressor

import numpy as np
import matplotlib.pyplot as plt

from sklearn import ensemble
from sklearn import datasets
from sklearn.utils import shuffle
from sklearn.metrics import mean_squared_error

In [20]:
params = {'n_estimators': 500, 'max_depth': 4, 'min_samples_split': 2,
          'learning_rate': 0.01, 'loss': 'ls'}
clf = ensemble.GradientBoostingRegressor(**params)

clf.fit(X_train, y_train)
mse = mean_squared_error(y_test, clf.predict(X_test))
print("MSE: %.4f" % mse)

In [24]:
print('MAE:', metrics.mean_absolute_error(y_test, clf.predict(X_test)))

from xgboost import XGBRegressor

MAE: 1271.08241976


Extreme Gradient boosting trees regression for Evaluation, prediction

In [25]:
# fitting model on training data

model = XGBRegressor(max_depth=6, n_estimators=500, learning_rate=0.1, subsample=0.8, colsample_bytree=0.4,
                     min_child_weight = 3,  seed=7)
model.fit(X_train, y_train)

XGBRegressor(base_score=0.5, colsample_bylevel=1, colsample_bytree=0.4,
       gamma=0, learning_rate=0.1, max_delta_step=0, max_depth=6,
       min_child_weight=3, missing=None, n_estimators=500, nthread=-1,
       objective='reg:linear', reg_alpha=0, reg_lambda=1,
       scale_pos_weight=1, seed=7, silent=True, subsample=0.8)

In [26]:
# Predictions on training and testing sets
pred_train = model.predict(X_train)
pred_test = model.predict(X_test)

from sklearn import metrics
print('MAE:', metrics.mean_absolute_error(y_test, pred_test))
print('MSE:', metrics.mean_squared_error(y_test, pred_test))
print('RMSE:', np.sqrt(metrics.mean_squared_error(y_test, pred_test)))

MAE: 1183.75313142
MSE: 3554715.15377
RMSE: 1885.39522482


In [16]:
#ensemble of models

from sklearn.datasets import load_boston
from sklearn.cross_validation import train_test_split
from sklearn.metrics import mean_absolute_error
from sklearn.ensemble import ExtraTreesRegressor
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor
from vecstack import stacking


X_train, X_test, y_train, y_test = train_test_split(X, Y, 
    test_size = 0.2, random_state = 0)


# Caution! All models and parameter values are just 
# demonstrational and shouldn't be considered as recommended.
# Initialize 1-st level models.

models = [
    ExtraTreesRegressor(random_state = 0, n_jobs = -1, 
        n_estimators = 100, max_depth = 3),
        
    RandomForestRegressor(random_state = 0, n_jobs = -1, 
        n_estimators = 100, max_depth = 3),
        
    XGBRegressor(seed = 0, nthread = -1, learning_rate = 0.1, 
        n_estimators = 100, max_depth = 3)]
    


In [47]:
    #S_train, S_test = stacking(models, X_train, y_train, X_test, 
   # regression = True, metric = mean_absolute_error, n_folds = 4, 
    #random_state = 0, verbose = 2)

Multilayer perceptron using keras and Tensorflow

In [4]:

import numpy
from pandas import read_csv
from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasRegressor
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline


Using TensorFlow backend.


importing the  training and test data

In [ ]:
df_train = pd.read_csv('train.csv')
df_test = pd.read_csv('test.csv')
print('training: ', df_train.shape)
print('test: ', df_test.shape)

Data preparation

In [5]:
#Convert to Numpy arrays and separate features/targets
training_samples = df_train.as_matrix()
training_targets = training_samples[:,-1]
training_samples = training_samples[:,1:-1]

test_samples = df_test.as_matrix()
test_samples = test_samples[:,1:]


#Encode the Labels of the categorical data
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
# [0:116]
allLabels = np.concatenate( ( training_samples[:, 0:116].flat , test_samples[:, 0:116].flat ) )
le.fit( allLabels )
del allLabels

#Transform the labels to int values
for colIndex in range(116):
    training_samples[:, colIndex] = le.transform(training_samples[:, colIndex])
    test_samples[:, colIndex] = le.transform( test_samples[:, colIndex] )

training:  (188318, 132)
test:  (125546, 131)


splitting the data into train and test

In [6]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(training_samples, training_targets)

Scaling the independent variables

In [7]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

/usr/local/lib/python3.5/dist-packages/sklearn/utils/validation.py:429: DataConversionWarning: Data with input dtype object was converted to float64 by StandardScaler.
  warnings.warn(msg, _DataConversionWarning)


imporitng the dependencies

In [19]:
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Dropout
from keras.wrappers.scikit_learn import KerasRegressor

Initiating the model:

In [20]:
Regressor =Sequential()

Regressor.add(Dense(output_dim = 66, init = 'uniform', activation ='relu', input_dim =130))

/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:3: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", kernel_initializer="uniform", units=66, input_dim=130)`
  This is separate from the ipykernel package so we can avoid doing imports until


In [21]:
Regressor.add(Dense(output_dim = 66, init = 'uniform', activation ='relu'))

/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:1: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", kernel_initializer="uniform", units=66)`
  """Entry point for launching an IPython kernel.


In [22]:
#output layer

Regressor.add(Dense(output_dim = 1, init = 'uniform', activation ='relu', input_dim =131))
Regressor.add(Dropout(p=0.1))

/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:3: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", kernel_initializer="uniform", units=1, input_dim=131)`
  This is separate from the ipykernel package so we can avoid doing imports until
/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:4: UserWarning: Update your `Dropout` call to the Keras 2 API: `Dropout(rate=0.1)`
  after removing the cwd from sys.path.


In [23]:
Regressor.compile(optimizer='adam',loss='mean_absolute_error')

In [26]:
Regressor.fit(X_train, y_train, batch_size =20, nb_epoch=50)

Epoch 1/50
   920/141238 [..............................] - ETA: 25s - loss: 1321.7657

/usr/local/lib/python3.5/dist-packages/keras/models.py:844: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.
  warnings.warn('The `nb_epoch` argument in `fit` '


141238/141238 [==============================] - 24s - loss: 1374.0053    
Epoch 2/50
141238/141238 [==============================] - 24s - loss: 1375.4119    
Epoch 3/50
141238/141238 [==============================] - 24s - loss: 1370.6883    
Epoch 4/50
141238/141238 [==============================] - 24s - loss: 1365.0149    
Epoch 5/50
141238/141238 [==============================] - 24s - loss: 1363.4576    
Epoch 6/50
141238/141238 [==============================] - 24s - loss: 1355.4415    
Epoch 7/50
141238/141238 [==============================] - 24s - loss: 1355.6239    
Epoch 8/50
141238/141238 [==============================] - 24s - loss: 1353.6335    
Epoch 9/50
141238/141238 [==============================] - 24s - loss: 1352.1357    
Epoch 10/50
141238/141238 [==============================] - 24s - loss: 1340.9816    
Epoch 11/50
141238/141238 [==============================] - 24s - loss: 1344.7981    
Epoch 12/50
141238/141238 [==============================] - 25

In [52]:
#Evaluating, Improving and Tuning the ANN

# Evaluating the ANN
from keras.wrappers.scikit_learn import KerasRegressor
from sklearn.model_selection import cross_val_score
from keras.models import Sequential
from keras.layers import Dense
def build_Regressor():
    Regressor = Sequential()
    Regressor.add(Dense(output_dim = 66, init = 'uniform', activation ='relu', input_dim =130))
    Regressor.add(Dense(output_dim = 66, init = 'uniform', activation ='relu'))
    Regressor.add(Dense(output_dim = 1, init = 'uniform', activation ='relu', input_dim =131))
    Regressor.compile(optimizer='adam',loss='mean_absolute_error')
    return Regressor
Regressor = KerasRegressor(build_fn = build_Regressor, batch_size = 10, epochs = 100)
accuracies = cross_val_score(estimator = Regressor, X = X_train, y = y_train, cv = 10, n_jobs = -1)
mean = accuracies.mean()
variance = accuracies.std()


/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:11: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", input_dim=130, kernel_initializer="uniform", units=66)`
  # This is added back by InteractiveShellApp.init_path()
/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:12: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(kernel_initializer="uniform", units=66, activation="relu")`
  if sys.path[0] == '':
/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:13: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", input_dim=131, kernel_initializer="uniform", units=1)`
  del sys.path[0]


Epoch 1/100


/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:11: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", input_dim=130, kernel_initializer="uniform", units=66)`
  # This is added back by InteractiveShellApp.init_path()
/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:12: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(kernel_initializer="uniform", units=66, activation="relu")`
  if sys.path[0] == '':
/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:13: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", input_dim=131, kernel_initializer="uniform", units=1)`
  del sys.path[0]


Epoch 1/100


/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:11: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", input_dim=130, kernel_initializer="uniform", units=66)`
  # This is added back by InteractiveShellApp.init_path()
/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:12: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(kernel_initializer="uniform", units=66, activation="relu")`
  if sys.path[0] == '':
/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:13: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", input_dim=131, kernel_initializer="uniform", units=1)`
  del sys.path[0]


Epoch 1/100


/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:11: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", input_dim=130, kernel_initializer="uniform", units=66)`
  # This is added back by InteractiveShellApp.init_path()
/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:12: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(kernel_initializer="uniform", units=66, activation="relu")`
  if sys.path[0] == '':
/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:13: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", input_dim=131, kernel_initializer="uniform", units=1)`
  del sys.path[0]


Epoch 1/100


KeyboardInterrupt: 

In [15]:
from keras.wrappers.scikit_learn import KerasRegressor
from sklearn.model_selection import GridSearchCV
from keras.models import Sequential
from keras.layers import Dense
def build_Regressor(optimizer):
    Regressor = Sequential()
    Regressor.add(Dense(output_dim = 66, init = 'uniform', activation ='relu', input_dim =130))
    Regressor.add(Dense(output_dim = 66, init = 'uniform', activation ='relu'))
    Regressor.add(Dense(output_dim = 1, init = 'uniform', activation ='relu', input_dim =131))
    Regressor.compile(optimizer='adam',loss='mean_absolute_error')
    return Regressor
Regressor = KerasRegressor(build_fn = build_Regressor)
parameters = {'batch_size': [25, 32],
              'epochs': [10, 15],
              'optimizer': ['adam', 'rmsprop']}
grid_search = GridSearchCV(estimator = Regressor,
                           param_grid = parameters,
                           cv = 10)
grid_search = grid_search.fit(X_train, y_train)
best_parameters = grid_search.best_params_
best_accuracy = grid_search.best_score_

/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:7: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", kernel_initializer="uniform", units=66, input_dim=130)`
  import sys
/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:8: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", kernel_initializer="uniform", units=66)`
  
/usr/local/lib/python3.5/dist-packages/ipykernel_launcher.py:9: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", kernel_initializer="uniform", units=1, input_dim=131)`
  if __name__ == '__main__':


Epoch 1/10
127114/127114 [==============================] - 16s - loss: 1370.9733    
Epoch 2/10
127114/127114 [==============================] - 16s - loss: 1230.5463    
Epoch 3/10
127114/127114 [==============================] - 16s - loss: 1219.7969    
Epoch 4/10
127114/127114 [==============================] - 16s - loss: 1212.6173    
Epoch 5/10
127114/127114 [==============================] - 16s - loss: 1207.2691    
Epoch 6/10
127114/127114 [==============================] - 16s - loss: 1203.3079    
Epoch 7/10
127114/127114 [==============================] - 16s - loss: 1199.5588    
Epoch 8/10
127114/127114 [==============================] - 16s - loss: 1196.3847    
Epoch 9/10
127114/127114 [==============================] - 16s - loss: 1193.6281    
Epoch 10/10
127114/127114 [==============================] - 16s - loss: 1383.3768    
Epoch 2/10
127114/127114 [==============================] - 16s - loss: 1234.4924    
Epoch 3/10
127114/127114 [===========================

KeyboardInterrupt: 

Model comparison: 
    
Linear Regression:  MAE: 1331.211
RIDGE  Regression:  MAE: 1331.230
lasso  Regression:  MAE: 1330.936
knn              :  MAE: 1715.262

decision trees    : MAE: 1745.644
decision trees 
with cross 
validation        : MAE: 1454.944
cart with 
parameter tuning  :MAE :1378.320
random forest     :MAE: 1375.474
adaboost          :MAE: 6840.072
gbm               :MAE: 1271.082
xgboost           :MAE: 1183.753
Multi layer
Perceptron         :MAE: 1192.468


    

Extreme gradient boosting algorithm is doing a better job in explaining the model.

The algorithms have been run without any fine tuning due to computational reasons.

The models use all the variables given in the data set and I'll be working to findout the best 
predictors and run further analysis on the data.

The Multilayer perceptron and the xgboost algorithms can be fine tuned to better the MAE.

Will be experimenting on how to stack models to gain a better accuracy in the next versions of this kernel